File: golang/src/p14/main.go
package main

import "fmt"

// n ->  n/2 (n is even)
// n -> 3n+1 (n is odd)
// sequence to 1 (e.g. 13, 40, 20, 10, 5, 16, 8, 4, 2, 1)
// Which starting number below 1 million has the most terms (10 above)
func main() {
	maxSeqLen := 0
	maxSeqLenNum := 0
	for i := 1; i < 1000000; i++ {
		l := seqLen(i)
		if l > maxSeqLen {
			maxSeqLen = l
			maxSeqLenNum = i
		}
	}

	fmt.Println(maxSeqLenNum)
}

func seqLen(n int) int {
	seqLen := 0
	for n != 1 {
		seqLen++
		n = next(n)
	}
	return seqLen
}

func next(n int) int {
	// in cpp and rust bit math vs mod+divide didn't seem to matter,
	// but in golang it does seem to make a difference.
	if n&1 == 1 {
		return 3*n + 1
	}
	return n >> 1

	//if n%2 == 1 {
	//	return 3*n + 1
	//}
	//return n / 2
}

File: cpp/src/p14/main.cc
#include <iostream>
#include <cstdint>

int next(uint32_t n) {
  // I thought this would be faster but it doesn't seem to matter.
  // Perhaps there's an optimization fixing it up.
  //return (n&1) ? (3*n+1) : (n>>1);
  return (n % 2 == 1)
            ? (3*n+1)
            : (n/2);
}

int seqLen(uint32_t n) {
  int seqLen = 0;
  while (n != 1) {
    seqLen++;
    n = next(n);
  }
  return seqLen;
}

int main() {
  int maxSeqLen = 0;
  int maxSeqLenNum = 0;
  for (uint32_t i = 1u; i < 1000000u; ++i) {
    const int l = seqLen(i);
    if (l > maxSeqLen) {
      maxSeqLen = l;
      maxSeqLenNum = i;
    }
  }
  std::cout << maxSeqLenNum << "\n";
}
